### PR TITLE
Property DDS: Minor fixes for TS 5.1

### PR DESCRIPTION
**Pull request by DLehenbauer**
This PR contains surgical fixes for 'property-changeset', 'property-properties', 'property-binder', and 'property-inspector-table' to unblock migration to the TypeScript 5 compiler.

### CODE REVIEW

File: experimental/PropertyDDS/packages/property-binder/src/data_binder/dataBinding.ts [hunk 1]
```diff
@@ -714,7 +714,9 @@ export class DataBinding {
 				typeidHolder,
 				in_indirectionsAtRoot,
 				in_previousSourceReferencePropertyInfo,
-				in_tokenizedFullPath.slice()!,
+				// WARNING: 'in_tokenizedFullPath' is '(string | number)[]'.  The cast to 'string[]'
+				//          preserves the JavaScript coercion behavior, which was permitted prior to TS5.
+				in_tokenizedFullPath.slice()! as string[],
 				tokenizedRegistrySubPath.slice()!,
```
Comment (DLehenbauer) [inline experimental/PropertyDDS/packages/property-binder/src/data_binder/dataBinding.ts:720]: The cast preserves the original behavior, but this one may be a bug?

File: experimental/PropertyDDS/packages/property-binder/src/data_binder/dataBinding.ts [hunk 2]
```diff
@@ -1106,7 +1108,13 @@ export class DataBinding {
...
 							break;
 						case ArrayChangeSetIterator.types.REMOVE:
-							for (i = 0; i < arrayIterator.opDescription.operation![1]; ++i) {
+							for (
+								i = 0;
+								// WARNING: 'operation[1]' is 'string | number | genericArray'.  The cast to 'number'
+								//          preserves the JavaScript coercion behavior, which was permitted prior to TS5.
+								i < (arrayIterator.opDescription.operation![1] as number);
+								++i
+							) {
```
Comment (DLehenbauer) [inline experimental/PropertyDDS/packages/property-binder/src/data_binder/dataBinding.ts:1117]: This cast preserves the original behavior.

Note that TypeScript 4 permitted implicit coercion to number for things like "5" < 10.
TypeScript 5 has reversed it's stance on this.

File: experimental/PropertyDDS/packages/property-inspector-table/src/ShareModal.tsx [hunk 3]
```diff
@@ -78,15 +78,25 @@ export interface IShareModalState {
...
+	) => Promise<any>;
 	/**
 	 * The handler that is invoked if the user decides to unshare.
 	 */
-	unshareHandler: (userIds: string[], groupIds: string[]) => Promise<any>;
+	unshareHandler: (
+		userIds: string[],
+		groupIds: string[],
+		permissions: { actions: SharingPermission[] },
+	) => Promise<any>;
```
Comment (DLehenbauer) [inline experimental/PropertyDDS/packages/property-inspector-table/src/ShareModal.tsx:99]: TypeScript 5 disallows passing an extra undeclared 3rd argument to the share/unshare callbacks.

I only found one reference to each callback, which passed '{ actions: ["read", "write", "delete"] }'.  I've added the necessary signature.

File: experimental/PropertyDDS/packages/property-properties/src/properties/abstractStaticCollectionProperty.js [hunk 4]
```diff
@@ -35,7 +35,7 @@ export class AbstractStaticCollectionProperty extends BaseProperty {
 	/**
 	 * Returns the sub-property having the given name, or following the given paths, in this property.
 	 *
-	 * @param {string|number|array<string|number>} in_ids - The ID or IDs of the property or an array of IDs if an array
+	 * @param {string | number | Array<string | number>} in_ids - The ID or IDs of the property or an array of IDs if an array
```
Comment (DLehenbauer) [inline experimental/PropertyDDS/packages/property-properties/src/properties/abstractStaticCollectionProperty.js:38]: When TypeScript cannot resolve the types declared in JSDoc comments it interprets these as 'unknown', some of which are starting to be caught by TypeScript 5's improved type checks.

(Here the problem is that 'Array' needs to be capitalized.)

File: experimental/PropertyDDS/packages/property-properties/src/properties/abstractStaticCollectionProperty.js [hunk 5]
```diff
@@ -347,10 +347,11 @@ export class AbstractStaticCollectionProperty extends BaseProperty {
 	 * Given an object that mirrors a PSet Template, assigns the properties to the values
 	 * found in that object.
 	 * See {@link setValues}
-	 * @param {object} in_values - The object containing the nested values to assign
+	 * @param {Object | Array | string} in_values - The object containing the nested values to assign
```
Comment (DLehenbauer) [inline experimental/PropertyDDS/packages/property-properties/src/properties/abstractStaticCollectionProperty.js:350]: Base class signatures must accept a superset of types accepted by overrides.

In this case, '_setValues' is overriden by MapProperty to accept an Object, ArrayProperty to accept an Array, and StringProperty to accept a string.

File: experimental/PropertyDDS/packages/property-properties/src/properties/abstractStaticCollectionProperty.js [hunk 6]
```diff
@@ -347,10 +347,11 @@ export class AbstractStaticCollectionProperty extends BaseProperty {
...
 	 * @param {boolean} in_typed - Whether the values are typed/polymorphic.
 	 * @param {boolean} in_initial - Whether we are setting default/initial values
 	 * or if the function is called directly with the values to set.
+	 * @protected
```
Comment (DLehenbauer) [inline experimental/PropertyDDS/packages/property-properties/src/properties/abstractStaticCollectionProperty.js:354]: TypeScript 5 notices if the visibility of an override does not match the base class signature.  In this case, '_setValues' was protected in the base class, but implicitly public here.

File: experimental/PropertyDDS/packages/property-properties/src/properties/baseProperty.ts [hunk 7]
```diff
@@ -589,9 +586,8 @@ export abstract class BaseProperty {
 	 * @param {property-properties.PathHelper.TOKEN_TYPES} in_segmentType - The type of segment in the tokenized path
 	 *
 	 * @return {property-properties.BaseProperty|undefined} The child property that has been resolved
-	 * @protected
 	 */
-	_resolvePathSegment(in_segment: string, in_segmentType: PathHelper.TOKEN_TYPES) {
+	protected _resolvePathSegment(in_segment: string, in_segmentType: PathHelper.TOKEN_TYPES) {
```
Comment (DLehenbauer) [inline experimental/PropertyDDS/packages/property-properties/src/properties/baseProperty.ts:590]: TypeScript ignores @protected JSDoc annotations for *.ts files.  Fixed by using the 'protected' keyword instead.

File: experimental/PropertyDDS/packages/property-properties/src/properties/baseProperty.ts [hunk 8]
```diff
@@ -670,7 +666,7 @@ export abstract class BaseProperty {
 	 *
 	 * @returns the flat representation
 	 */
-	private _flatten(): object {
+	protected _flatten(): object {
 		return { propertyNode: this };
```
Comment (DLehenbauer) [inline experimental/PropertyDDS/packages/property-properties/src/properties/baseProperty.ts:670]: TypeScript 5 noticed a private method was colliding with another private method.

In this case, I think the intention was to override, and so I promoted '_flatten' to protected in both cases.

File: experimental/PropertyDDS/packages/property-properties/src/properties/baseProperty.ts [hunk 9]
```diff
@@ -689,10 +685,8 @@ export abstract class BaseProperty {
 
 	/**
 	 * Return a JSON representation of the properties and its children.
-	 * @return {object} A JSON representation of the properties and its children.
-	 * @private
 	 */
-	_toJson() {
+	protected _toJson(): Object {
```
Comment (DLehenbauer) [inline experimental/PropertyDDS/packages/property-properties/src/properties/baseProperty.ts:689]: Ditto: I think the intention was to override _toJson() on subclasses.

**Review by taylorsw04 — APPROVED** [review]

**Comment by DLehenbauer** [comment]
Thanks, @taylorsw04 !

**Pull request merged**